### ISSUE TITLE
Linting ideas

### ISSUE THREAD

**Issue by boris-kolpackov** [build2/build2#197]
We've been thinking about having a `buildfile` (or, more globally, project) linter or linting mode. We already have some checks like this spread out throughout the toolchain (for example, making sure that separate tests/examples/benchmark packages belong to the same project as the main package). We can start capturing some linting ideas in this issue.

- Detect utility libraries (`libu?{}`) as prerequisites of `./` (this is unnecessary and in case of `libul{}` can mess up for-install'ness).

- Detect directory symlinks and make sure they are listed as such in `.gitattributes` ([required for `git` to work correctly on Windows](https://build2.org/article/symlinks.xhtml#windows)).

**Comment by Klaim** [comment]
Some ideas inspired by some recent common issues I had when learning `build2` and new users I'm helping hit:

- Detect when `build/export.build` uses paths that refer to non-existing directories. This is classic when you renamed a project/package (for example when switching from 1package repository to N packages repository), renamed directories and you forgot to update the `export.build`.

- Detect all broken symlinks. This is silencious and lead to indirectly related errors, which are hard to understand to beginners.

- Detect when there is a submodule but it's not initialized.

This is a classic issue when you just cloned a project and forgot to add `--recursive` or didn't `git submodule update --recursive`. In that situation all the behavior of `build2` looks unusual and doesn't clearly point to the issue, but `build2` could probably emit a warning about that by checking if there is a submodule declared in `.gitsubmodule` but not currently initialized (the directory doesnt exist).

**Comment by boris-kolpackov** [comment]
Here is an important one:

* Detect conditional dependencies:

  ```
  if ($cxx.target.class != 'windows')
    ./: exe{posix-only-test}
  ```
  
  Which should be:
  
  ```
  ./: exe{posix-only-test}: include = ($cxx.target.class != 'windows')
  ```
  
In fact, I've wasted so much time pointing this mistake out in submitted packages (that fail to build due to missing source files) that I wonder if maybe we should make it a hard error during normal parse.

**Comment by Klaim** [comment]
> In fact, I've wasted so much time pointing this mistake out in submitted packages (that fail to build due to missing source files) that I wonder if maybe we should make it a hard error during normal parse.

Indeed: to me the first code "looks" legitimate (it makes sense while reading and won't trigger my "smell" detector even if knowing there might be an issue) while the second code looks like a shortcut for the first code (kind of like a ternary if).
If the first code leads to issues, making it a hard error would definitely help adopt the right way to do it, in particular if the error message hints at the right way to do it.

**Comment by boris-kolpackov** [comment]
Ok, for 0.15.0, I've added this check as a warning.

**Comment by boris-kolpackov** [comment]
Here is another candidate: check that `{c,cxx}.?options` are not specified on `lib{}` or `exe{}` targets. For some reason this appears to be a frequent mis-expectation (maybe because of the similar semantics in GNU `make`; we don't do the same since there could be multiple, say, libraries that depend on the same prerequisite which brings the question: which `*.?options` are we going to take -- maybe also add a note in the manual with the rationale).

This would have to be a rule-specific check which makes me wonder if perhaps `lint` should (also) be a meta-operation, similar to `configure` and `dist`.
